### PR TITLE
Implement secure boost scheme phase 1 - vertical pipeline with hist sync

### PR DESCRIPTION
**Pull request by ZiyueXu77**
For implementing Vertical Federated Learning with Secure Features, as discussed in 
https://github.com/dmlc/xgboost/issues/9987 
The first phase is to implement an alternative vertical pipeline that sync the histograms from clients to label owner.
This PR implemented this feature as a standalone data mode.

Functional changes finished, currently adding unit testings

Note: phase 2 will be adding HE encryption features, which will be added in an independent PR

### CODE REVIEW

**Comment by rongou** [comment]
@trivialfis

File: include/xgboost/data.h
Comment (rongou) [inline include/xgboost/data.h:43]: I'm not sure this is really a data split mode. Would we ever need encryption for horizontal split?

Comment (ZiyueXu77) [reply include/xgboost/data.h:43]: for now, horizontal split does not need encryption. but potentially if HE supports further computation, then yes we could also add encryption to horizontal. This datasplit mode leads to below line 189-193 
" bool IsColumnSplit() const { return (data_split_mode == DataSplitMode::kCol) || (data_split_mode == DataSplitMode::kColS); }

  /** @brief Whether the data is split column-wise with secure computation. */
  bool IsSecure() const { return data_split_mode == DataSplitMode::kColS; }
" 
Maybe in the future we can call it "data mode" instead of "data split mode"?

File: src/common/quantile.cc
Comment (trivialfis) [inline src/common/quantile.cc:396]: Why 0?

File: src/common/quantile.cc
Comment (trivialfis) [inline src/common/quantile.cc:374]: where is this used? also, `1e-5` is not the minimum feature value, XGBoost can take all real number left and right to 0.

File: src/common/quantile.cc
Comment (trivialfis) [inline src/common/quantile.cc:482]: wouldn't it leak feature values to other workers?

File: src/tree/hist/evaluate_splits.h
Comment (trivialfis) [inline src/tree/hist/evaluate_splits.h:352]: so rank 0 is the label owner, hence the "server"?

**Review by trivialfis — COMMENTED** [review]
Thank you for working on federated learning! Exciting new features.

Initial review as I'm still reading the secure boost paper. It would be great if you could make a summary of the various differences between all data modes.

File: src/common/quantile.cc
Comment (ZiyueXu77) [reply src/common/quantile.cc]: removed

File: src/common/quantile.cc
Comment (ZiyueXu77) [reply src/common/quantile.cc:395]: good point， this part is to be used with the below "cut values need to be synced" part, will address with new inference pattern

File: src/tree/hist/evaluate_splits.h
Comment (ZiyueXu77) [reply src/tree/hist/evaluate_splits.h:352]: yes, rank 0 is considered label owner / active party / server

File: src/common/quantile.cc
Comment (trivialfis) [inline src/common/quantile.cc:381]: We are filling it with 0, not the minimum value. Please update the comment.

File: src/common/quantile.cc
Comment (trivialfis) [inline src/common/quantile.cc:399]: Is it possible that we can reuse or merge the `AddCutPoint` function somehow? Seems not ideal to copy code.

File: src/tree/hist/evaluate_splits.h
Comment (trivialfis) [inline src/tree/hist/evaluate_splits.h]: please cleanup the code.

File: src/tree/hist/evaluate_splits.h
Comment (trivialfis) [inline src/tree/hist/evaluate_splits.h:414]: Is this part of the code even useful for passive parties? Considering that they don't evaluate splits. If not, then it would be much cleaner to skip the call to evaluation altogether. Keep spreading conditions like `if (secure)` can make the code difficult to change.

File: src/tree/hist/histogram.h
Comment (trivialfis) [inline src/tree/hist/histogram.h:196]: Why do we need to allgather the histogram across workers? I thought we only need to send it to the active worker?

File: src/tree/hist/histogram.h
Comment (trivialfis) [inline src/tree/hist/histogram.h:199]: what's the issue with the current system? I'm working on the network module, maybe I can help?

File: src/tree/hist/evaluate_splits.h
Comment (ZiyueXu77) [reply src/tree/hist/evaluate_splits.h]: removed

File: src/tree/hist/histogram.h
Comment (ZiyueXu77) [reply src/tree/hist/histogram.h:196]: yes we only need to collect histograms to the active party, but my understanding is we currently do not have a "gather" function to do that?

File: src/tree/hist/histogram.h
Comment (ZiyueXu77) [reply src/tree/hist/histogram.h:199]: sorry for the confusion, what I mean is that AllReduce is the most direct and effective solution to handle this computation (all histograms have the same length, those "empty" slots hold value 0, so AllReduce can update them in place).

File: src/common/quantile.cc
Comment (ZiyueXu77) [reply src/common/quantile.cc:403]: To be handled by another PR

File: src/common/quantile.cc
Comment (ZiyueXu77) [reply src/common/quantile.cc:489]: to be handled by another PR

File: src/common/quantile.cc
Comment (ZiyueXu77) [reply src/common/quantile.cc:381]: thanks, updated

File: src/common/quantile.cc
Comment (ZiyueXu77) [reply src/common/quantile.cc:399]: good point

File: src/tree/hist/evaluate_splits.h
Comment (ZiyueXu77) [reply src/tree/hist/evaluate_splits.h:414]: currently the (secure && passive parties) is skipped with "if ((!is_secure_) || (collective::GetRank() == 0)) {", recommendations on skipping it in other places?

File: src/tree/hist/histogram.h
Comment (ZiyueXu77) [reply src/tree/hist/histogram.h]: comment updated

File: src/tree/hist/histogram.h
Comment (trivialfis) [reply src/tree/hist/histogram.h:196]: Thank you for sharing, I can look into a gather function in the future.

**Comment by trivialfis** [comment]
I have enabled all the CI pipelines, please don't push until they finish, otherwise a new commit will interrupt the previous run. The PR looks good to me overall and will approve once all tests pass.

Please note that after having all the desired features in the feature branch and having a full picture of the code changes, we might do a few rounds of refactors before merging into the master. This way we can unblock these individual PRs while keeping the code maintainable in the future.

**Comment by ZiyueXu77** [comment]
> I have enabled all the CI pipelines, please don't push until they finish, otherwise a new commit will interrupt the previous run. The PR looks good to me overall and will approve once all tests pass.
> 
> Please note that after having all the desired features in the feature branch and having a full picture of the code changes, we might do a few rounds of refactors before merging into the master. This way we can unblock these individual PRs while keeping the code maintainable in the future.

sounds good! Thanks a lot. :)

**Review by trivialfis — APPROVED** [review]

**Pull request merged**
